### ISSUE TITLE
Dynamic resource references

### ISSUE THREAD

**Issue by a-hilaly** [aws-controllers-k8s/community#1576]
The ACK ecosystem currently supports referencing resources from the same or different GVKs within a static implementation. For example, you can reference an `s3::bucket` in a `lambda::function` object. However, this implementation does not allow for the referencing of multiple resource types.

Ideally, we would like to expand this feature to support the referencing of any object managed by ACK controllers. This would allow for resources like `AWS EventBridge` Rule targets to reference a wide range of resources, such as `sns::topic` and `sqs::queue`.

The ACK ecosystem's current ability to reference resources is a new feature that was developed in response to a recent problem. @embano1 and I will continue exploring and working on a design document for this feature, which will involve refactoring the original resource reference implementation to allow for the referencing of multiple resource types. This will require moving the generation logic to the code-generator/pkg/generate package and switching to the k8s dynamic client.

**Comment by embano1** [comment]
+1  to simplify the code base (eventually) using dynamic client and make it easier to compose resources using `TypedLocalObjectReferences`.

For example, as an `EventBridge` controller user I want to reference an `SQS` queue as a target which is also managed by ACK:

```yaml
apiVersion: eventbridge.services.k8s.aws/v1alpha1
kind: Rule
metadata:
  name: rule-01
  namespace: namespace-01
spec:
  name: rule-01
  eventPattern: "{\"detail-type\":\"testevent\"}"
  # keep as is
  eventBusRef:
    name: eventbus-01
  # allows by arn and TypedLocalObjectReference
  targets:
    - id: someResourceFromArn
      arn: arn:aws:1234567890:someresource
    - id: queueFromRef
      objectRef:
        apiGroup: sqs.services.k8s.aws/v1alpha1
        kind: Queue
        name: queue-01
```

To be discussed: do we allow cross-namespace resource references (`ObjectReference`) or resources must be within the same namespace? I can see use cases for both.

**Comment by jaypipes** [comment]
@A-Hilaly @embano1 @RedbackThomson @jljaco I think it's probably a good idea to just use the dynamic client anyway for everything... that way we can remove the need to include in the `go.mod` file the referenced controller libs.

**Comment by embano1** [comment]
Yup, will help with a lot of imports and deps! Kinda feels like generics :-p

**Comment by jljaco** [comment]
After some discussion, @RedbackThomson and I would like to see a more full-fledged design proposal for this feature before we can potentially slate it for future work.

@A-Hilaly @jaypipes @embano1 anyone have bandwidth to come up with a more detailed proposal?

**Comment by embano1** [comment]
Happy to assist as EventBridge is going to benefit a lot from this and I brought this up with @A-Hilaly based on my experience with Knative, but can't sign up for owning this AI.

**Comment by jaypipes** [comment]
> After some discussion, @RedbackThomson and I would like to see a more full-fledged design proposal for this feature before we can potentially slate it for future work.
> 
> @A-Hilaly @jaypipes @embano1 anyone have bandwidth to come up with a more detailed proposal?

@jljaco apologies, I don't have bandwidth to work on this one right now

**Comment by a-hilaly** [comment]
@alyIbrahim @kejne @joestadler @migueleliasweb We just open sourced `kro`, [a new composition engine for Kubernetes](https://github.com/awslabs/kro), which can be used with ACK as well as any custom resource definitions. `kro` solves the problem of dynamic references, so you won't need to use ACK references at all. :) Just instruct kro to query status fields and inject them into the spec of other resources.

**Comment by migueleliasweb** [comment]
Yeah, I just saw it actually. Thanks @a-hilaly . It looks sweet! I was just discussing this with my team!

Looks similar to https://docs.crossplane.io/latest/concepts/compositions/ (I'm sure there was some inspiration there :sweat_smile:)

Do you think moving forward ACK will be ported more directly to inside `kro` or will it be kept its own thing and just used within the `compositions`?

**Comment by a-hilaly** [comment]
>Do you think moving forward ACK will be ported more directly to inside kro or will it be kept its own thing and just used within the compositions?

`kro` focuses exclusively on Kubernetes resource orchestration and composition by interacting only with the Kubernetes API. ACK and kro will continue to operate as independent projects and can be used together if needed. `kro` will maintain its fundamentally generic design, ensuring full compatibility with ANY Kubernetes resource (native and custom) across all the environments.
